### PR TITLE
refactor: simplify NFTView container

### PR DESCRIPTION
**Pull request by joonchoi0918**
[Associated Notion Card](https://www.notion.so/Refactor-NFTView-component-8c119c46cd274464b42b61d452da4596)

-----

## 1. Pull request checklist

- [x] Notion card has been moved to the Code Review column
- [x] Notion card has a link to this PR
- [x] A reviewer has been assigned to the Notion card

### CODE REVIEW

File: src/containers/other/NFTView/hooks/useNftData.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	const highestBidAmount: number | undefined = useMemo(() => {
+		return highestBid?.amount
+			? Number(ethers.utils.formatEther(highestBid.amount))
+			: undefined;
+	}, [highestBid]);
+
+	/**
+	 * Callback functions
+	 *
+	 */
```
Comment (ericlowturing) [inline src/containers/other/NFTView/hooks/useNftData.ts:86]: 👍

File: src/containers/other/NFTView/hooks/useNftData.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+			try {
+				const events = (await sdk.instance?.getDomainEvents(
+					znsDomain.id,
+				)) as DomainEvents[];
+				setHistories(sortHistoriesByTimestamp(events));
+			} catch (e) {
+				console.error('Failed to retrieve bid data');
+			} finally {
+				setIsHistoryLoading(false);
+			}
```
Comment (ericlowturing) [inline src/containers/other/NFTView/hooks/useNftData.ts:100]: Oh! 💯

File: src/containers/other/NFTView/hooks/useNftData.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+				}
+			}
+
+			setHighestBid(highestBid);
+			setBuyNowPrice(Number(ethers.utils.formatEther(buyNow)));
+		} catch (e) {
+			console.error('Failed to retrieve price data');
+		} finally {
+			setIsPriceDataLoading(false);
+		}
```
Comment (ericlowturing) [inline src/containers/other/NFTView/hooks/useNftData.ts:149]: 👍

File: src/containers/other/NFTView/hooks/useNftData.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	const shareAsset = useCallback(async () => {
+		if (domain) {
+			shareDomainAsset(domain);
+		}
+	}, [domain]);
+
+	/**
+	 * Life cycle
+	 *
+	 */
```
Comment (ericlowturing) [inline src/containers/other/NFTView/hooks/useNftData.ts:171]: 👍

File: src/containers/other/NFTView/NFTView.tsx [hunk 5]
```diff
@@ -1,632 +1,105 @@
...
-import NFT from './components/NFT';
-import HistoryItem from './components/HistoryItem';
+import {
+	NFT,
+	Stats,
+	TokenHashBoxes,
+	Attributes,
+	Histories,
+	DomainSettings,
+} from './elements';
```
Comment (ericlowturing) [inline src/containers/other/NFTView/NFTView.tsx:18]: Oh this is BIG change of large component. Nice working @joonchoi0918

File: src/containers/other/NFTView/NFTView.types.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+	DomainMintEvent,
+	DomainBidEvent,
+	DomainSaleEvent,
+} from '@zero-tech/zns-sdk/lib/types';
+
+export type DomainEvents =
+	| DomainTransferEvent
+	| DomainMintEvent
+	| DomainBidEvent
+	| DomainSaleEvent;
```
Comment (ericlowturing) [inline src/containers/other/NFTView/NFTView.types.ts:12]: Nice union types! Btw there is not this type defined in SDK? @cc @colbr

**Review by ericlowturing — COMMENTED** [review]
LGTM. Nice refactoring. I left a few comments @joonchoi0918

File: src/containers/other/NFTView/NFTView.helpers.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+						reject();
+					}
+				});
+			});
+		};
+
+		try {
+			const asset = await Promise.any([
+				checkUrl(NFT_ASSET_URLS.VIDEO.replace(/NFT_ASSET_HASH/g, hash)),
+				checkUrl(NFT_ASSET_URLS.IMAGE.replace(/NFT_ASSET_HASH/g, hash)),
```
[the remark this answers is not in the document]
Comment (colbr) [reply src/containers/other/NFTView/NFTView.helpers.ts:57]: Agreed, I like this more than the string concatenation we were doing before

File: src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+					<b>
+						3
+						<a
+							className="alt-link"
+							href={`https://etherscan.io/address/${history.seller!}`}
+							target="_blank"
+							rel="noreferrer"
+						>{`${history.seller!.substring(0, 4)}...${history.seller!.substring(
+							history.seller!.length - 4,
+						)}`}</a>
```
[the remark this answers is not in the document]
Comment (colbr) [reply src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx:116]: There's a library function called `truncateAddress` which should serve this purpose

File: src/containers/other/NFTView/NFTView.types.ts (hunk 6 above)
Comment (colbr) [reply src/containers/other/NFTView/NFTView.types.ts:12]: I don't think so! At least not in the current version

**Review by colbr — CHANGES_REQUESTED** [review]
Looking really nice, good job. The only required changes are the `Histories` ones I mentioned in feedback :)

I'm also about to push a few changes to `NFTView.tsx` in `develop` which will probably give you some conflicts - I'm happy to resolve them

File: src/config.ts [hunk 9]
```diff
@@ -1,5 +1,6 @@
 const config = {
 	env: process.env.ENVIRONMENT,
+	baseURL: process.env.REACT_APP_ZNS_SHARE_BASE_URL,
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/config.ts:3]: Yeah, it works as I expected.

File: src/containers/other/NFTView/NFTView.constants.tsx [hunk 10]
```diff
@@ -1,5 +1,48 @@
-export const MORE_ACTION_KEY = {
-	MY_DOMAIN_SETTINGS: 'My Domain Settings',
-	VIEW_BIDS: 'View Bids',
-	TRANSFER_OWNERSHIP: 'Transfer Ownership',
+//- Asset Imports
+import settingsIcon from './assets/settings.svg';
+import transferOwnershipIcon from './assets/transfer.svg';
+import dollarSignIcon from './assets/dollar-sign.svg';
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/NFTView.constants.tsx:4]: No worries. It works 👍

File: src/containers/other/NFTView/NFTView.helpers.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+};
+
+export const getDomainAsset = async (
+	znsDomain: Maybe<DisplayParentDomain>,
+): Promise<string | undefined> => {
+	if (znsDomain?.animation_url || znsDomain?.image_full || znsDomain?.image) {
+		// Get hash from asset
+		const url = (znsDomain.animation_url ||
+			znsDomain.image_full ||
+			znsDomain.image)!;
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/NFTView.helpers.ts:39]: Will fix it.

File: src/containers/other/NFTView/NFTView.helpers.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/NFTView.helpers.ts:57]: I prefer to use REGEX and it makes code cleaner

File: src/containers/other/NFTView/elements/Attributes/Attributes.tsx [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+						</li>
+					))}
+
+					{/* Show / Hide more button */}
+					{initialHiddenAttributesCount > 0 && (
+						<button
+							className={`${styles.ToggleAttributes} ${
+								!isCollapsed && styles.SetOpacityAnimation
+							}`}
+							style={{ background: 'none' }}
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Attributes/Attributes.tsx:82]: I just copied the JSX from old NFTView.tsx, let me update scss and remove inline styles

File: src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+		const history = item as DomainSaleEvent;
+
+		return (
+			<li className={styles.Bid}>
+				<div>
+					<b>
+						3
+						<a
+							className="alt-link"
+							href={`https://etherscan.io/address/${history.seller!}`}
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx:111]: Ok, constants! Will fix it

File: src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx:116]: Yeah, I just moved this component folder. Will look and fix it. Thanks

File: src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+import { ethers } from 'ethers';
+import { DomainEvents } from '../../../NFTView.types';
+import styles from '../../../NFTView.module.scss';
+
+type HistoryItemProps = {
+	item: DomainEvents;
+};
+
+const HistoryItem = ({ item }: HistoryItemProps) => {
+	if (item.type === DomainEventType.bid) {
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx:18]: Agreed! Switch Case!

File: src/containers/other/NFTView/elements/Stats/Stats.tsx [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+			try {
+				const domainMetricsCollection: DomainMetricsCollection =
+					await sdk.instance.getDomainMetrics([znsDomain.id]);
+				setDomainMetrics(domainMetricsCollection[znsDomain.id]);
+				setIsDomainMetricsLoaded(true);
+			} catch (e) {
+				console.error(e);
+			}
+		}
+	}, [sdk, znsDomain]);
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Stats/Stats.tsx:65]: No, It is not dependency

File: src/containers/other/NFTView/elements/Stats/Stats.tsx [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+
+	useEffect(() => {
+		if (znsDomain) {
+			fetchStats();
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [znsDomain]);
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Stats/Stats.tsx:72]: I don't agree. It does not work in this case

File: src/containers/other/NFTView/NFTView.helpers.ts [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+export const truncateText = (
+	text: string,
+	startLength: number,
+	endLength: number = 4,
+): string => {
+	return `${text.slice(0, startLength)}...${text.slice(
+		text.length - endLength,
+	)}`;
+};
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/NFTView.helpers.ts:30]: Yeah, I created it

File: src/containers/other/NFTView/NFTView.helpers.ts [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+	return bids.sort((a, b) =>
+		asc
+			? Number(ethers.utils.formatEther(b.amount)) -
+			  Number(ethers.utils.formatEther(a.amount))
+			: Number(ethers.utils.formatEther(a.amount)) -
+			  Number(ethers.utils.formatEther(b.amount)),
+	);
+};
+
+export const sortHistoriesByTimestamp = (
```
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/NFTView.helpers.ts:124]: Will look and fix it. Thanks

[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/NFTView.helpers.ts]: @colbr I have updated

File: src/containers/other/NFTView/NFTView.helpers.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/NFTView.helpers.ts]: Done

File: src/containers/other/NFTView/elements/Attributes/Attributes.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Attributes/Attributes.tsx]: Done

File: src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx]: Done

File: src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx]: Done

File: src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (joonchoi0918) [reply src/containers/other/NFTView/elements/Histories/elements/HistoryItem.tsx]: Done

**Review by colbr — APPROVED** [review]

**Comment by colbr** [comment]
Looking good, thanks for making the requested changes! Will hold off on merging this for a couple of days so we can get the major release out. I will handle any merge conflicts that pop up

**Comment by colbr** [comment]
Pointing this PR to another branch to merge with some other refactoring, then will merge to `develop` when ready 😊

**Pull request merged**